substrate/primitives/trie/src/hasher_random_state.rs

```diff
@@ -0,0 +1,171 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utility module to use a custom random state for HashMap and friends
+//! in a no_std environment.
+
+use core::{
+	cell::UnsafeCell,
+	hash::Hasher as CoreHasher,
+	sync::atomic::{AtomicU8, Ordering},
+};
+
+use core::hash::BuildHasher;
+use hashbrown::DefaultHashBuilder;
+
+// Constants to represent the state of the global extra randomness.
+// UNINITIALIZED: The extra randomness has not been set yet.
+const UNINITIALIZED: u8 = 0;
+// LOCKED: The extra randomness is being set.
+const LOCKED: u8 = 1;
+// INITIALIZED: The extra randomness has been set and is ready to use.
+const INITIALIZED: u8 = 2;
+
+// SAFETY: we only mutate the UnsafeCells when state is in the thread-exclusive
+// LOCKED state, and only read when state is in the INITIALIZED state.
+unsafe impl Sync for GlobalExtraRandomnesss {}
+struct GlobalExtraRandomnesss {
+	initialized: AtomicU8,
+	randomness: UnsafeCell<[u8; 16]>,
+}
+
+// Extra randomness to be used besides the one provided by the `DefaultHashBuilder`.
+static EXTRA_RANDOMNESS: GlobalExtraRandomnesss = GlobalExtraRandomnesss {
+	initialized: AtomicU8::new(UNINITIALIZED),
```

Comment:
I don't think we will have any kind of threading here soon. But also not a problem here.